Add MapDataset.from_images

AtreyeeS opened the pull request:
<!-- These are hidden commments. You can leave them or delete them. -->

<!-- Don't be shy! Github issues or pull requests are welcome any time. -->
<!-- But if you first want to chat, use the Gammapy Slack. -->
<!-- See https://gammapy.org/contact.html -->

**Description**
<!-- What is the motivation for this pull request? -->
<!-- Briefly: what changes are done here? -->
<!-- If this is to address a Github issue, metion it's number to create a link -->
<!-- Example: This PR is to fix issue #42 -->

This pull request addresses #2913 

**Dear reviewer**
<!-- Let the reviewer and Gammapy team know what you want: -->
<!-- * Is this ready for review? Or is it work in progress and you want some feedback? -->
<!-- * Do you want ot go through review here? Or if someone just finish this up and merge it in? -->
<!-- Do you have any specific questions, e.g. about API or implementation? -->
<!-- Do you include a test executing new code you're adding (to make sure it runs)? -->
<!-- Do you include some documentation? Is it needed? -->

registerrier commented on gammapy/datasets/map.py:
```diff
@@ -1128,6 +1134,59 @@ def from_dict(cls, data, lazy=False, cache=True):
 
+    @classmethod
+    def from_images(cls, images, energy_min, energy_max, name=None, **kwargs):
+        """Create a MapDataset from a set 2D images which do not have an
```
from a dictionary of

registerrier commented on gammapy/datasets/map.py:
```diff
@@ -1128,6 +1134,59 @@ def from_dict(cls, data, lazy=False, cache=True):
+        Parameters
+        ----------
+        images: dict
+        Dictionary of images. Allowed keywords are `counts`, `background`,
```
need to increase indentation here

registerrier commented on gammapy/datasets/map.py:
```diff
@@ -1893,6 +1959,57 @@ def from_geoms(
+            images, energy_min, energy_max, name=None, **kwargs
+        )
+
+        energy_axis = MapAxis.from_energy_edges([energy_min, energy_max], name="energy")
```
you can retrieve it from the maps contained in `dataset`

registerrier commented on gammapy/datasets/map.py:
```diff
@@ -1128,6 +1134,59 @@ def from_dict(cls, data, lazy=False, cache=True):
+        kwargs["name"] = name
+
+        energy_axis = MapAxis.from_energy_edges([energy_min, energy_max], name="energy")
+        energy_axis_true = MapAxis.from_energy_edges(
```
What if the user has a different binning for psf kernel and exposure?

adonath replied:
But we always require that the PSF kernel and exposure have the same energy axis, except if the PSF is in reco energy. Is that what you meant?

registerrier reviewed (commented):
Thanks @AtreyeeS .

This looks good. But I have a problem with the energy axis definition. Here the only supported case is when energy and energy true axes are identical. Isn't it an issue?

AtreyeeS commented:
Hi @registerrier : Actually I thought about this as well - Then it would need an energy dispersion kernel. Can that normally be available without having an energy axis on the exposure?

adonath commented on gammapy/datasets/map.py:
```diff
@@ -1128,6 +1134,59 @@ def from_dict(cls, data, lazy=False, cache=True):
+
+        Parameters
+        ----------
+        images: dict
```
Maybe the documentation and usage becomes simpler (especially for the `PSFKernel`  case) if the maps are passed directly like: `.from_images(counts, background, exposure, psf_kernel,  )`. I think there is also some documentation missing...

AtreyeeS commented:
I removed the lines from the main code base and added them as an example in the H.E.S.S. notebook. More than think should be as internal documentation I guess.

adonath commented:
Thanks @AtreyeeS for the update. There are a two things I'm not yet happy with:
- I think we should not advertise to use the `MapDatasetOnOff` for spatial fitting, even within HESS I think this has never been done. Instead just use `MapDataset` with the `BACKGROUNDMAP`
- The way the exposure is currently computed from `EXPGAMMAMAP` is incorrect I think. To get correct flux values for fitting  of models, I think one needs to (a) define the energy axis with `e_min = 1 TeV` and `e_max = np.inf * u.TeV` (or at  least a very large number...) and (b) always assume a power-law with spectral index of 2.3 or whatever index  has been chosen respectively.

AtreyeeS commented:
So do you propose
- add a `datasetonoff.to_mapdataset()` before going for fitting, or directly creating a `MapDataset`? I would slightly prefer the former.
- Energy range can be changed, but there is no Spectral Model involved at this stage. Where do I incorporate this?

adonath commented:
> add a datasetonoff.to_mapdataset() before going for fitting, or directly creating a MapDataset? I would slightly prefer the former.

I think I would directly create a `MapDataset` here, because the only thing one could do with the `MapDatasetOnOff` is to use the `ExcessMapEstimator`, however all the maps computed by the `ExcessMapEstimator` are already present in the HAP maps. So I think the only fair use case it fitting... 

> Energy range can be changed, but there is no Spectral Model involved at this stage. Where do I incorporate this?

I think this has to remain a comment / note at this stage...

registerrier commented on docs/tutorials/data/hess.ipynb:
```diff
@@ -12,7 +12,17 @@
     "\n",
```
<p>unnecessary</p><p> </p><p>Do we really want to put the Note written like this? After all we should have the opportunity to introduce data at the DL4 level.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3350/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

registerrier replied:
<p>Do we need this? Would it really work like this?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3350/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

registerrier replied:
<p>as explained</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/gammapy/gammapy/pull/3350/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

adonath commented:
@AtreyeeS, what do we do about this PR?

adonath commented:
Just discussed in the dev call: right now there are no public HESS maps we could use to document e.g. a morphology fit. So the proposal is to move the documentation to internal HESS documentation instead. If there is public data at some point we can add some public documentation to Gammapy as well...
